snapcraft/models/project.py

```diff
@@ -2358,106 +2416,89 @@ class Core24Project(Project):
-    base: SkipJsonSchema[None] = None  # type: ignore[assignment,reportIncompatibleVariableOverride]
-    build_base: Literal["core24"]  # type: ignore[reportIncompatibleVariableOverride]
+        We're overriding this because pydantic does not allow before validators on
+        discriminator fields.
+        """
+        return build_base
 
 
-class _BuildBaseProjectEnum(Enum):
-    DEVEL = "devel"
-    CORE22 = "core22"
-    CORE24 = "core24"
-    UNIMPLEMENTED = "UNIMPLEMENTED"
+class Core26Project(Core24Project):
+    base: Literal["core26", "devel"]  # type: ignore[assignment]
+    build_base: Literal["devel"]  # pyright: ignore[reportGeneralTypeIssues,reportIncompatibleVariableOverride]
 
-    @classmethod
-    def _missing_(cls, value: Any):
-        return cls.UNIMPLEMENTED
+    grade: Annotated[  # type: ignore[reportIncompatibleVariableOverride]
+        Literal["devel"],
+        _custom_error("grade must be 'devel' when build-base is 'devel'"),
+    ]
 
 
-class _BaseProjectEnum(Enum):
-    CORE22 = "core22"
-    CORE24 = "core24"
-    BARE = "bare"
-    UNIMPLEMENTED = "UNIMPLEMENTED"
+class BareCore26Project(Core26Project):
+    base: Literal["bare"]  # type: ignore[assignment]
+    build_base: Literal["devel"]  # type: ignore[assignment]
 
+    @override
+    @pydantic.field_validator("build_base", mode="after")
     @classmethod
-    def _missing_(cls, value: Any):
-        return cls.UNIMPLEMENTED
-
-
-class ProjectType(Enum):
-    BASE = "base"
-    APP = "app"
-    GADGET = "gadget"
-    KERNEL = "kernel"
-    SNAPD = "snapd"
-    NONE = "None"
-
+    def _validate_devel_base(
+        cls, build_base: str, info: pydantic.ValidationInfo
+    ) -> str:
+        """Override _validate_devel_base from craft-application to be a no-op.
 
-def _discriminator(enum: type[Enum], key: str):
-    kebab_cased_key = key.replace("_", "-")
-    return lambda data: enum(data.get(key, data.get(kebab_cased_key))).value
+        We're overriding this because pydantic does not allow before validators on
+        discriminator fields.
+        """
+        return build_base
 
 
-def _type_discriminator(data: dict):
-    snap_type = data.get("type")
+_BareProject = Annotated[
+    BareCore22Project | BareCore24Project | BareCore26Project,
+    pydantic.Discriminator("build_base"),
+]
 
-    if snap_type not in ProjectType._value2member_map_:
-        snap_type = ProjectType.NONE.value
+_StandardProject = Annotated[
+    Core22Project | Core24Project | Core26Project | _BareProject,
+    pydantic.Discriminator("base"),
+]
 
-    return snap_type
 
+def _validate_type(data: dict[str, Any]) -> dict[str, Any]:
+    if (
+        data.get("base") is None  # Catch both missing base key and null base.
+        and data.get("type") not in ("base", "kernel", "snapd")
+    ):
+        raise ValueError(
+            "Snap base must be declared when type is not base, kernel or snapd"
```

Comment:
```suggestion
            "You must set the 'base' key when the snap type is not 'base', 'kernel' or 'snapd'"
```

Reply:
```suggestion
            f"Missing 'base' key for snap type {data.get('type')!r}"
```

Reply:
Thoughts @medubelko ? This is phrased as an error but I believe makes the remedy clear (either add a base or change the type).

It's basically your first suggestion but with fancy string handling. An example output would be:

`Missing 'base' key for snap type 'app'.`

(If you're okay with this I'll need to further refine the code to take care of some edge cases such as type being undefined and the fact that I forgot the fullstop, but I'm going to wait until I hear back on whether that dynamism is good or bad.)